library/src/main/scala/scala/scalajs/js/annotation/internal/JSOptional.scala

```diff
@@ -0,0 +1,19 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2016, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package scala.scalajs.js.annotation.internal
+
+import scala.annotation.meta._
+
+/** IMPLEMENTATION DETAIL: Marks a member of a JavaScript trait as an optional
+ *  field, that must be emitted.
+ *
+ *  Do not use this annotation yourself.
+ */
+@field @getter @setter
+class JSOptional extends scala.annotation.StaticAnnotation
```

Comment:
> I think we are missing tests (and code) that this annotation cannot be used in source code (like we have for other internal annotations).

Indeed.

> Now that this is internal, I'm wondering if this is the best name (the other extreme of the name spectrum "granularity" would be @JSIgnore). I have been thinking about this for a while, but haven't been able to come up with a fully satisfying name.

Yeah, I wasn't very happy with the name either, but couldn't come with better. It's quite technical :s